core/sr-primitives/src/weights.rs

```diff
@@ -74,3 +82,139 @@ impl Default for TransactionWeight {
 		TransactionWeight::Basic(0, 1)
 	}
 }
+
+/// A wrapper for fee multiplier.
+/// This is to simulate a `Perbill` in the range [-1, infinity]
+///
+/// The fee multiplier is always multiplied by the weight (as denoted by `TransactionWeight` on a
+/// per-transaction basis with `#[weight]` annotation) of the transaction to obtain the final fee.
+///
+/// One can define how this conversion evolves based on the previous block weight by implementing
+/// the `FeeMultiplierUpdate` type of the `system` trait.
+#[cfg_attr(feature = "std", derive(PartialEq, Eq, Debug))]
+#[derive(Encode, Decode, Clone, Copy)]
+pub enum FeeMultiplier {
+	/// Should be interpreted as a positive ratio added to the weight, i.e. `weight + weight * p`
+	/// where `p` is a small `Perbill`.
+	///
+	Positive(Perbill, Weight),
+	/// Should be interpreted as a negative ratio subtracted from the weight, i.e.
+	/// `weight - weight * p` where `p` is a small `Perbill`.
+	Negative(Perbill),
+}
+
+impl FeeMultiplier {
+	/// Applies the self, as a multiplier, to the given weight.
+	pub fn apply_to(&self, weight: Weight) -> Weight {
+		match *self {
+			FeeMultiplier::Positive(p, r) => weight + weight.saturating_mul(r).saturating_add(p * weight),
```

Comment:
maybe I am confused but it feels like we're adding an extra `weight` term here. This essentially reduces to `weight + weight * r + weight * p`, while I thought we wanted to return `weight + weight * p`. I expected something closer to
```suggestion
			FeeMultiplier::Positive(p, weight) => weight.saturating_add(weight.saturating_mul(p)),
```

Reply:
Correct concern but it is correct; note that there is a `1 + ..` in the formula. Also, note how this does not exist for the negative portion.